File: sdk/ruby/lib/thetabase/scribe.rb
# frozen_string_literal: true

# The host half of Scribe: sockets, and nothing else.
#
# Every decision about the protocol — how a request is encoded, what a response
# means, when a cached read must be dropped — lives in the WebAssembly core
# (`crates/theta-scribe-wasm`). This file moves bytes.
#
# The split is why a seventh SDK is cheap. A Ruby SDK that built wire messages in
# Ruby would be a seventh implementation of the protocol, and the first time the
# schema moved six of the seven would be wrong. What must be identical across
# languages is the protocol; what must differ is I/O.

require "json"
require "socket"
require "wasmtime"

module ThetaBase
  # A protocol error the core rejected, kept distinct from a transport failure.
  #
  # The distinction is the whole point: a caller may retry a transport failure
  # and must not retry a protocol one.
  class ProtocolError < StandardError; end

  # The WebAssembly core, loaded once per process.
  class Scribe
    # `[u32 length][u8 tag][bytes]`, as `abi.rs` lays it out.
    RESULT_HEADER = 5
    LENGTH_PREFIX_BYTES = 4

    # What the core returns for a frame length this protocol does not allow.
    NO_LENGTH = 0xffffffff

    def initialize(wasm)
      engine = Wasmtime::Engine.new
      # No host imports: the core is pure computation, which is why it loads
      # identically here, in Node, in a browser, in Go, on the JVM and in .NET.
      @store = Wasmtime::Store.new(engine)
      module_ = Wasmtime::Module.new(engine, wasm)
      @instance = Wasmtime::Instance.new(@store, module_)
      @memory = @instance.export("memory")&.to_memory
      raise ProtocolError, "this Scribe core exports no memory" if @memory.nil?

      @mutex = Mutex.new
    end

    def self.load(path)
      new(File.binread(path))
    end

    # ---- the core's surface -------------------------------------------------

    # Frame a request.
    def encode(request, request_id, branch_id)
      call("theta_encode", JSON.generate(request), request_id, branch_id)
    end

    # Read a response body into a document.
    def decode(body)
      JSON.parse(call("theta_decode", body))
    end

    # Read the four-byte frame prefix.
    def body_length(prefix)
      @mutex.synchronize do
        pointer = invoke("theta_alloc", LENGTH_PREFIX_BYTES)
        begin
          @memory.write(pointer, prefix)
          length = invoke("theta_body_length", pointer) & 0xffffffff
          if length == NO_LENGTH
            raise ProtocolError, "the peer sent a length this protocol does not allow"
          end

          length
        ensure
          invoke("theta_free", pointer, LENGTH_PREFIX_BYTES)
        end
      end
    end

    # Frame the handshake.
    #
    # The protocol version is the core's to state, not the host's: a host that
    # could name its own version could claim a compatibility it does not have,
    # and the negotiation that refuses mismatched versions rather than guessing
    # would be negotiating with itself.
    def encode_hello(token, client_name)
      call("theta_encode_hello", JSON.generate({ token: token, clientName: client_name }))
    end

    # Read the server's reply to a handshake. A refusal raises.
    def decode_welcome(body)
      JSON.parse(call("theta_decode_welcome", body))
    end

    # Render a typed query AST to SQL-subset source and bound parameters.
    #
    # In the core rather than here, so a query built the same way in Python
    # renders to the same bytes — and so the rule that a value never becomes
    # query text has one implementation.
    def render_query(ast)
      JSON.parse(call("theta_render_query", JSON.generate(ast)))
    end

    # What a request invalidates: one key, everything, or nothing.
    def invalidation(request)
      JSON.parse(call("theta_invalidation", JSON.generate(request)))
    end

    private

    def invoke(name, *args)
      export = @instance.export(name)&.to_func
      # Named rather than deferred to a nil error at the first call: a core built
      # from a different revision is a real failure mode, and
      # "theta_render_query is missing" says which revision.
      raise ProtocolError, "#{name} is not exported by this Scribe core" if export.nil?

      export.call(*args)
    end

    def call(name, input, *extra)
      # The core owns one linear memory and every call allocates into it, so two
      # threads encoding at once would interleave allocations against one heap.
      @mutex.synchronize do
        bytes = input.dup.force_encoding(Encoding::BINARY)
        pointer = invoke("theta_alloc", bytes.bytesize)
        begin
          @memory.write(pointer, bytes)
          take(invoke(name, pointer, bytes.bytesize, *extra))
        ensure
          invoke("theta_free", pointer, bytes.bytesize)
        end
      end
    end

    # Read a result buffer out of the core's memory and free it.
    #
    # The tag byte distinguishes a result from a refusal, and it is checked
    # before the payload is used — the refusal path must not depend on the
    # success path having been valid.
    def take(pointer)
      header = @memory.read(pointer, RESULT_HEADER)
      length = header[0, 4].unpack1("V")
      tag = header.getbyte(4)
      payload = @memory.read(pointer + RESULT_HEADER, length)
      invoke("theta_free_result", pointer)

      raise ProtocolError, payload.force_encoding(Encoding::UTF_8) unless tag.zero?

      payload.force_encoding(Encoding::UTF_8)
    end
  end

  # A framed connection to `thetad`.
  class Connection
    def initialize(core, socket)
      @core = core
      @socket = socket
    end

    def write(bytes)
      @socket.write(bytes)
    end

    # Read exactly +n+ bytes, or raise if the peer closes first.
    def read_fully(n)
      # A socket delivers whatever arrived, not what was asked for, and a short
      # read here would frame the next message wrong.
      out = @socket.read(n)
      raise IOError, "the server closed the connection" if out.nil? || out.bytesize < n

      out
    end

    # Read one length-prefixed frame.
    def frame
      read_fully(@core.body_length(read_fully(Scribe::LENGTH_PREFIX_BYTES)))
    end

    # One request/response exchange.
    #
    # Sequential by construction: a caller that needs concurrency opens more
    # connections. Multiplexing on one socket would need request-id correlation
    # on the read side, and correlating replies to the wrong request is a
    # data-corruption bug rather than a performance one.
    def exchange(request, request_id, branch_id)
      write(@core.encode(request, request_id, branch_id))
      @core.decode(frame)
    end

    def close
      @socket.close
    end
  end
end

File: sdk/python/src/thetabase/scribe.py
"""The host half of Scribe: sockets, and nothing else.

Every decision about the protocol — how a request is encoded, what a response
means, when a cached read must be dropped — lives in the WebAssembly core
(``crates/theta-scribe-wasm``). This module moves bytes.

The split is deliberate. A Python SDK that built wire messages in Python would
be a second implementation of the protocol and the TypeScript one a third, and
the first time the schema moved two of the three would be wrong. What must be
identical across languages is the protocol; what must differ is I/O, because
WebAssembly has no sockets and the host runtimes do not agree on what one is.
"""

from __future__ import annotations

import json
import socket as _socket
from pathlib import Path
from typing import Any

# ``[u32 length][u8 tag][bytes]``, as ``abi.rs`` lays it out.
_RESULT_HEADER = 5
LENGTH_PREFIX_BYTES = 4


def _compact(value: Any) -> bytes:
    """JSON with no incidental whitespace.

    ``json.dumps`` pads with spaces and ``JSON.stringify`` does not, so without
    this the Python and TypeScript hosts hand the core different bytes for the
    same request. Nothing on the wire changes — the core re-encodes into Cap'n
    Proto either way — but two hosts that disagree about what they send are two
    hosts whose behaviour can diverge somewhere it does matter.
    """
    return json.dumps(value, separators=(",", ":")).encode("utf-8")


class ProtocolError(Exception):
    """A protocol error the core rejected, distinct from a transport failure."""


class ScribeCore:
    """The WebAssembly core, loaded once per process."""

    def __init__(self, wasm_path: str | Path | None = None) -> None:
        # Imported here rather than at module scope: the type surface in
        # ``__init__`` is usable without a WebAssembly runtime installed, and a
        # caller who only wants the types should not be made to install one.
        import wasmtime

        path = Path(wasm_path) if wasm_path else _default_module()
        self._store = wasmtime.Store()
        module = wasmtime.Module.from_file(self._store.engine, str(path))
        # No imports: the core is pure computation, which is why the same module
        # loads in Python, Node, a browser and an edge runtime.
        self._instance = wasmtime.Instance(self._store, module, [])
        self._memory = self._instance.exports(self._store)["memory"]

    def _fn(self, name: str) -> Any:
        return self._instance.exports(self._store)[name]

    def _write(self, data: bytes) -> int:
        ptr = self._fn("theta_alloc")(self._store, len(data))
        self._memory.write(self._store, data, ptr)
        return ptr

    def _take(self, ptr: int) -> bytes:
        header = self._memory.read(self._store, ptr, ptr + _RESULT_HEADER)
        length = int.from_bytes(header[:4], "little")
        tag = header[4]
        body = bytes(
            self._memory.read(
                self._store, ptr + _RESULT_HEADER, ptr + _RESULT_HEADER + length
            )
        )
        self._fn("theta_free_result")(self._store, ptr)

        if tag != 0:
            raise ProtocolError(body.decode("utf-8", "replace"))
        return body

    def _call(self, name: str, data: bytes, *args: Any) -> bytes:
        ptr = self._write(data)
        try:
            return self._take(self._fn(name)(self._store, ptr, len(data), *args))
        finally:
            self._fn("theta_free")(self._store, ptr, len(data))

    def encode(self, request: dict[str, Any], request_id: int, branch_id: int) -> bytes:
        """A host request as a framed wire message."""
        return self._call("theta_encode", _compact(request), request_id, branch_id)

    def decode(self, body: bytes) -> dict[str, Any]:
        """A wire response as JSON."""
        return json.loads(self._call("theta_decode", body).decode("utf-8"))

    def body_length(self, prefix: bytes) -> int:
        """How many bytes of body follow a length prefix."""
        ptr = self._write(prefix)
        try:
            length = self._fn("theta_body_length")(self._store, ptr)
        finally:
            self._fn("theta_free")(self._store, ptr, len(prefix))

        if length == 0xFFFFFFFF:
            raise ProtocolError("the peer sent a length this protocol does not allow")
        return int(length)

    def encode_hello(self, token: str, client_name: str) -> bytes:
        """Encode the handshake.

        The protocol version is the core's to state, not the host's: a host that
        could name its own version could claim a compatibility it does not have,
        and the negotiation that refuses mismatched versions rather than
        guessing would be negotiating with itself.
        """
        return self._call(
            "theta_encode_hello", _compact({"token": token, "clientName": client_name})
        )

    def decode_welcome(self, body: bytes) -> dict[str, Any]:
        """Decode the server's reply to a handshake. A refusal raises."""
        return json.loads(self._call("theta_decode_welcome", body).decode("utf-8"))

    def render_query(self, ast: dict[str, Any]) -> dict[str, Any]:
        """Render a typed query AST to SQL-subset source and bound parameters.

        In the core rather than here, so a query built the same way in
        TypeScript renders to the same bytes — and so the rule that a value
        never becomes query text has one implementation.
        """
        return json.loads(self._call("theta_render_query", _compact(ast)).decode("utf-8"))

    def invalidation(self, request: dict[str, Any]) -> dict[str, Any]:
        """What this request invalidates: one key, everything, or nothing."""
        return json.loads(self._call("theta_invalidation", _compact(request)).decode("utf-8"))


def _default_module() -> Path:
    """Find the core next to the package, then in the build tree."""
    # `here` is `sdk/python/src/thetabase`, so the repository root is
    # `parents[3]`. It was `parents[4]` — one level above the checkout — which
    # meant the build-tree fallback never resolved and the only way to load the
    # core from a checkout was to pass its path by hand. Nothing caught it
    # because the conformance runners do exactly that.
    here = Path(__file__).resolve().parent
    candidates = [
        # Beside the package: how it is laid out once installed.
        here / "theta_scribe_wasm.wasm",
        here.parents[2] / "theta_scribe_wasm.wasm",
        # In the build tree: how it is laid out in a checkout.
        here.parents[3] / "target/wasm32-unknown-unknown/wasm/theta_scribe_wasm.wasm",
    ]
    for candidate in candidates:
        if candidate.exists():
            return candidate
    raise FileNotFoundError(
        "the Scribe core was not found — run `make wasm` to build it, or pass its "
        "path to ScribeCore()"
    )


class Connection:
    """One framed connection to ``thetad``.

    Sequential by construction: a caller that needs concurrency opens more
    connections. Multiplexing one socket would need request-id correlation on
    the read side, and correlating a reply to the wrong request is a
    data-corruption bug rather than a slow one.
    """

    def __init__(self, core: ScribeCore, sock: _socket.socket) -> None:
        self._core = core
        self._sock = sock
        self._next_id = 1

    def _read_exactly(self, count: int) -> bytes:
        chunks = []
        remaining = count
        while remaining:
            chunk = self._sock.recv(remaining)
            if not chunk:
                raise ConnectionError(
                    f"the peer closed with {remaining} of {count} bytes still to come"
                )
            chunks.append(chunk)
            remaining -= len(chunk)
        return b"".join(chunks)

    def send_raw(self, framed: bytes) -> bytes:
        """Write a framed message and read one framed reply back."""
        self._sock.sendall(framed)
        prefix = self._read_exactly(LENGTH_PREFIX_BYTES)
        return self._read_exactly(self._core.body_length(prefix))

    def call(self, request: dict[str, Any], branch_id: int = 0) -> dict[str, Any]:
        request_id = self._next_id
        self._next_id += 1
        body = self.send_raw(self._core.encode(request, request_id, branch_id))
        return self._core.decode(body)

    def close(self) -> None:
        self._sock.close()
